onnxruntime/core/mlas/lib/wasm/sgemmv.cpp

```diff
@@ -0,0 +1,171 @@
+/*++
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    sgemmc.cpp
+
+Abstract:
+
+    This module implements the kernels for the single precision matrix/matrix
+    multiply operation (SGEMM).
+
+--*/
+
+#include "mlasi.h"
+
+size_t
+MLASCALL 
+MlasGemvFloatKernel(
+    const float* A,
+    const float* B,
+    float* C,
+    size_t CountK,
+    size_t CountN,
+    size_t ldb,
+    bool ZeroMode
+    )
+/*++
+
+Routine Description:
+
+    This routine is an inner kernel to compute matrix multiplication for a
+    set of rows. This handles the special case of M=1.
+
+    The elements in matrix B are not transposed.
+
+Arguments:
+
+    A - Supplies the address of matrix A.
+
+    B - Supplies the address of matrix B.
+
+    C - Supplies the address of matrix C.
+
+    CountK - Supplies the number of columns from matrix A and the number
+        of rows from matrix B to iterate over.
+
+    CountN - Supplies the number of columns from matrix B and matrix C to
+        iterate over.
+
+    ldb - Supplies the first dimension of matrix B.
+
+    ZeroMode - Supplies true if the output matrix must be zero initialized,
+        else false if the output matrix is accumulated into.
+
+Return Value:
+
+    None.
+
+--*/
+{
+    if (ZeroMode && CountK > 0) {
+        float* c = C;
+        const float* b = B;
+        const float A0 = A[0];
+        auto N = CountN;
+        constexpr size_t kWidth = 4;
+        for (; N >= kWidth; N -= kWidth) {
+            c[0] = A0 * b[0];
+            c[1] = A0 * b[1];
+            c[2] = A0 * b[2];
+            c[3] = A0 * b[3];
+            c += kWidth;
+            b += kWidth;
+        }
+
+        for (; N > 0; N--) {
+            c[0] = A0 * b[0];
+            c++;
+            b++;
+        }
+        A++;
```

Comment:
will it mess up the simd alignment?

Reply:
oh, this is for wasm without SIMD. Then what's the benefit to compute 4 elements together?

---
In reply to: [599117576](https://github.com/microsoft/onnxruntime/pull/6478#discussion_r599117576) [](ancestors = 599117576)

Reply:
unroll will help decrease the loop count.